### ISSUE TITLE
Reduce runtime for Citus regression tests

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#346]
(I'm copy / pasting our email thread to continue the conversation here)

I was checking the regression tests to reduce their running time.

Current Running Time
\----------
I run check-multi and check-full on my computer (4 core 1.8 GHz).
* Running time of check-multi: 125 seconds (~10 seconds for initialization and shut-down)
* Running time of check-full: 450 seconds

Critical Tests in Check-Multi
\----------
Here are the slowest tests and their approximate times during check-multi:
* multi_subquery: 5 sec
* multi_single_relation_subquery: 15 sec
* parallel group of multi_prepare_sql and multi_prepare_plsql: 40 sec 
* multi_large_table_join_planning : 7 sec
* multi_large_table_pruning: 7 sec
* multi_large_table_task_assignment: 10 sec

Speeding Up Check-Multi
\----------
To improve check-multi, I need to make more tests run parallel, especially the slowest ones. Here there are two possible ideas:

1. Trying to make the slowest tests parallel without touching their content if they are not dependent
2. Breaking the slow tests into smaller pieces and run these pieces in parallel

To try the first idea, I put the the multi_single_relation_subquery and the parallel group (multi_prepare_sql and multi_prepare_plsql) together because these are the slowest ones and not dependent to each other. However, it only decreases the total time approximately 5 seconds. Also, it seems there are not too many options for this kind of improvement but I need to investigate a bit more.

To try the second idea, I divided the tests of the parallel group (multi_prepare_sql and multi_prepare_plsql) into smaller pieces and put them parallel. It decreases the total time approximately 15 seconds and it seems easy to find more options for this kind of improvement. Intuitively, I am waiting to decrease the total time of check-multi to 80-90 seconds with better organization and by dividing the more tests. However, I am suspicious about its maintenance because I needed to create 10 more tests for this specific example.

Speeding Up Check-Full
\----------
In check-full, we run the test suit of check-multi three times so improving check-multi will improve check-full, too. Also, we don't cover too many new parameters by re-running the suit because of explicit settings in the tests. Here, Andres suggested to replace the check-multi-binary with additional explicit tests. This will remove unnecessary testing and the initialization time of check-multi-binary (~approximately 10 sec for each suit). To remove initialization time of other suits, we can also integrate check-worker and check-multi-fdw into the main schedule.

These changes may decrease time of check-full to approximately 300 seconds without the improvements on check-multi.

**Comment by mtuncer** [comment]
When we need to see certain log message like being router plannable/executable, or to see pruned shards we set client min messages to DEBUG2. It spits lots of log messages that contains job and shard ids.

When this test is run in parallel, we get different results depending on how the tests are scheduled, therefore end up removing failing test from parallel group.

We observe this behavior when we stage data or create new shards. I wonder if we can group such tasks together and have shard creation done per test group. This could let us increasing parallelism.

**Comment by mtuncer** [comment]
Following tests explicitly set task executor type, therefore no need to run them in task-tracker mode one more time.
```
multi_binary_master_copy_format
multi_complex_count_distinct
multi_explain
multi_large_table_join_planning
multi_large_table_pruning
multi_large_table_task_assignment
multi_prepare_plsql
multi_prepare_sql
multi_repartitioned_subquery_udf
multi_router_planner
multi_simple_relation_subquery
multi_subquery
```

**Comment by mtuncer** [comment]
So this is what I did
1. add 2 more schedules for running check-full instead of running check-multi. These schedules are stripped down versions of check-multi, that do not any more value when run again.
2. explicitly set task-tracker-delay and remote-task-check-interval.

I found this has reduced runtime of both check-multi and check-full to about half on my laptop.

**Comment by mtuncer** [comment]
I only made 1 test to run in parallel (multi_complex_count_distinct).  Since we reduced process waiting times, making test to run in parallel did not yield expected performance benefits. 

With final changes, check-multi completes less than 1/3 of the original time on my laptop.
check full is completes between 1/3 to half of the original time.  Right now we have fix test setup costs for each test run. 

I think this is enough performance improvements until next time we feel the need.
